### ISSUE TITLE
Europa: docker.#Run doesn’t work with docker.#Build

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#1446]
## Problem

Two major actions in the new `universe.dagger.io/docker` don’t work together: `docker.#Run` and `docker.#Build`. It is extremely common to run containers as part of a container build, so this is a major issue that breaks virtually all other packages in some way.

In short:

* `docker.#Build` defines an interface for implementing build steps. The interface is `input?: #Image, output: #Image`.
* `docker.#Run` cannot cleanly implement this interface as-is, for 2 reasons:
  * `input: #Image` would be redundant with its existing field `image: #Image`
  * `output: #Image` would conflict with its existing field `output: { directories: {…}, files: {…} }`
* A workaround was attempted that involves hacking the two interfaces together. However it does not work, and consequently `main` does not allow using `#Run` and `#Build` together.

## Solution

*TLDR: option 4 is the favorite, there is one more detail to discuss and we are ready to merge #1438.*

Several solutions have been discussed:

[Option 1](#1434): move build stuff to a docker/build package, with aliases for everything except #Run which would be split.
Option 2: take over #Run.input and #Run.output for build step interface, move the rest to a new export field
Option 3: change build step interface to build: input and build: output. This allows #Run to be more explicit in supporting build, but forces every other step to be more explicit too.
[Option 4](https://github.com/dagger/dagger/pull/1438): same as option 2, but support either `input:` or `image:` with a disjunction.

### Option 1 (spin out `docker/build`)

Pros:
* Cleanest implementation
* Simplest interface for `docker.#Run`
Cons:
* `build.#Build` is really awkward to use
* Possible issues of drift. Do I use `build.#Run` or `docker.#Run`? What about other build steps?

### Option 4 (“friendly takeover”)

In this option the build API takes over `input` and `output`, but the takeover is “friendly” because we still get to use `image:` when outside of a build.

Pros:
* Keep everything in one package
* `docker.#Build` and `docker.#Run` just make sense
* `export:` is arguably even better than `output:`

Cons:
* `output` must be optional, similar to `export.rootfs` (formerly `output.rootfs`). This is because not every run requires exporting the root filesystem. Before merging, we would need to check if this pattern is affected.
